internal/pkg/earpointer/analysis.go

```diff
@@ -0,0 +1,649 @@
+	case *ssa.Next:
+		// Next is handled in visitExtract().
+	case *ssa.Return:
+		// Return is handled in visitCall().
+	case *ssa.RunDefers:
+		// To be implemented.
+	case *ssa.Panic:
+		// To be implemented.
+	case *ssa.DebugRef:
+		// Not related.
+	}
+}
+
+func (vis *visitor) visitFieldAddr(addr *ssa.FieldAddr) {
+	if !typeMayShareObject(addr.Type()) {
+		return
+	}
+	// "t1 = t0.x" is implemented by t0[x -> t1] through address unification.
+	obj := addr.X
+	field := makeField(obj.Type(), addr.Field)
+	for _, c := range vis.contexts {
+		vis.unifyFieldAddress(c, obj, field, addr)
+	}
+}
+
+func (vis *visitor) visitField(field *ssa.Field) {
+	// "t1 = t0.x" is implemented by t0[x -> t1].
+	obj := field.X
+	fd := makeField(obj.Type(), field.Field)
+	vis.processHeapAccess(field, obj, *fd)
+}
+
+func (vis *visitor) visitIndexAddr(addr *ssa.IndexAddr) {
+	// "t2 = &t1[t0]" is handled through address unification.
+	vis.processMemoryIndexAccess(addr, addr.X, addr.Index)
+}
+
+func (vis *visitor) visitIndex(index *ssa.Index) {
+	vis.processMemoryIndexAccess(index, index.X, index.Index)
+}
+
+func (vis *visitor) visitLookup(lookup *ssa.Lookup) {
+	if !lookup.CommaOk {
+		vis.processMemoryIndexAccess(lookup, lookup.X, lookup.Index)
+	}
+	// The comma case will be processed by the subsequent extract instruction.
+}
+
+func (vis *visitor) visitStore(store *ssa.Store) {
+	// Here store instruction "*address = data" is implemented by a
+	// semantics-equivalent (w.r.t. unification) load instruction "data =
+	// *address".
+	// The address can be a (memory) address, a global variable, or a free
+	// variable.
+	if mayShareObject(store.Val) {
+		vis.processAddressToValue(store.Addr, store.Val)
+	}
+}
+
+func (vis *visitor) visitPhi(phi *ssa.Phi) {
+	if !typeMayShareObject(phi.Type()) {
+		return
+	}
+	for _, e := range phi.Edges {
+		vis.unifyLocals(phi, e)
+	}
+}
+
+func (vis *visitor) visitMapUpdate(update *ssa.MapUpdate) {
+	vis.processMemoryIndexAccess(update.Value, update.Map, update.Key)
+}
+
+func (vis *visitor) visitConvert(convert *ssa.Convert) {
+	vis.unifyLocals(convert, convert.X)
+}
+
+func (vis *visitor) visitChangeInterface(change *ssa.ChangeInterface) {
+	vis.unifyLocals(change, change.X)
+}
+
+func (vis *visitor) visitChangeType(change *ssa.ChangeType) {
+	vis.unifyLocals(change, change.X)
+}
+
+func (vis *visitor) visitUnOp(op *ssa.UnOp) {
+	switch op.Op {
+	case token.ARROW: // channel read
+		// Use array read to simulate channel read.
+		vis.processHeapAccess(op, op.X, anyIndexField)
+	case token.MUL:
+		// v = *addr
+		// The LHS is a register; the RHS can be a (memory) address, a global
+		// variable, or a free variable.
+		// t1 = *addr is implemented as "addr[directPointToField] = t1",
+		// i.e., address addr points to value t1.
+		vis.processAddressToValue(op.X, op)
+	}
+}
+
+func (vis *visitor) visitBinOp(op *ssa.BinOp) {
+	// Numeric binop will be skipped.
+	// The only binop that matters now is "+" that concatenates strings.
```

Comment:
Is the type of `op.X` / `op.Y` knowable at this moment?  If we only are considering strings here, it would be nice to declare that in code and not just comment.

I don't expect any real-world false positives without it, so if it's not feasible, no big deal.